controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -86,25 +85,49 @@ public String getScopeName() {
      * @return : future of whether it was done or not
      */
     @Override
-    public CompletableFuture<Boolean> create(final StreamConfiguration configuration, long createTimestamp) {
-        final Create create = new Create(createTimestamp, configuration);
+    public CompletableFuture<CreateStreamResponse> create(final StreamConfiguration configuration, long createTimestamp) {
 
         return checkScopeExists()
-                .thenCompose(x -> checkStreamExists(create))
-                .thenCompose(x -> storeCreationTime(create))
-                .thenCompose(x -> createConfiguration(create))
-                .thenCompose(x -> createState(State.CREATING))
-                .thenCompose(x -> createSegmentTable(create))
-                .thenCompose(x -> createNewEpoch(0))
-                .thenCompose(x -> {
-                    final int numSegments = create.getConfiguration().getScalingPolicy().getMinNumSegments();
-                    final byte[] historyTable = TableHelper.createHistoryTable(create.getCreationTime(),
-                            IntStream.range(0, numSegments).boxed().collect(Collectors.toList()));
+                .thenCompose(x -> checkStreamExists(configuration, createTimestamp))
+                .thenCompose(createStreamResponse -> storeCreationTimeIfNotExists(createStreamResponse.getTimestamp())
+                        .thenCompose(x -> createConfigurationIfNotExists(createStreamResponse.getConfiguration()))
+                        .thenCompose(x -> createStateIfNotExists(State.CREATING))
+                        .thenCompose(x -> createNewSegmentTable(createStreamResponse.getConfiguration(), createStreamResponse.getTimestamp()))
+                        .thenCompose(x -> getState().thenCompose(state -> {
+                            if (state.equals(State.CREATING)) {
+                                return createNewEpoch(0);
```

Comment:
epoch 0 is created at the time of stream creation. 
If a stream is in creating state, we will try to idempotently create epoch 0.
if its a previously created stream which is already in "Active" state, it may have potentially gone through several scale operations hence may have newer epochs.. if we dont have this check, we may end up creating epoch 0 along with epoch N.. though active epoch will still be returned from the history table which will be N.. but having epoch 0 present in the metadata would be incorrect.